src/commands/update.rs

```diff
@@ -0,0 +1,236 @@
+use clap::ArgMatches;
+use log::{debug, error, info};
+
+use std::{fs, io::ErrorKind, path::Path, process::exit};
+
+use crate::{
+  constants, files::config::load_config, server, steamcmd::steamcmd_command, utils::get_working_dir,
+};
+
+enum Action {
+  Check,
+  Force,
+  Regular,
+}
+
+impl Action {
+  fn new(check: bool, force: bool) -> Self {
+    match (check, force) {
+      (true, true) => panic!("`check` and `force` are mutually exlusive!"),
+      (true, false) => Self::Check,
+      (false, true) => Self::Force,
+      (false, false) => Self::Regular,
+    }
+  }
+}
+
+pub fn invoke(args: &ArgMatches) {
+  info!("Checking for updates");
+
+  if !server::is_installed() {
+    error!("Failed to find server executable. Can't update!");
+    exit(1);
+  }
+
+  let dry_run = args.is_present("dry_run");
+  let check = args.is_present("check");
+  let force = args.is_present("force");
+
+  let server_running = server::is_running();
+  let update_available = update_is_available();
+  if update_available {
+    info!("A server update is available!");
+  } else {
+    info!("No server updates found");
+  }
+
+  match Action::new(check, force) {
+    Action::Check => match (dry_run, update_available) {
+      (true, true) => {
+        info!("Dry run: An update is available. This would exit with 0 to indicate this.")
+      }
+      (true, false) => {
+        info!("Dry run: No update is available. This would exit with 1 to indicate this.")
+      }
+      // 0 indicates that an update is available
+      (false, true) => exit(0),
+      // TODO: should we do a value other than 1 here, and if we do then what value?
+      // 1 indicates the server is up to date
+      (false, false) => exit(1),
+    },
+    Action::Force => match (dry_run, server_running) {
+      (true, true) => info!("Dry run: Server would be shutdown, updated, and brought back online"),
+      (true, false) => info!("Dry run: The server is offline and would be updated"),
+      _ => {
+        debug!("Force updating!");
+        update_server();
+      }
+    },
+    Action::Regular => {
+      if dry_run {
+        match (server_running, update_available) {
+          (true, true) => info!(
+            "Dry run: An update is available and the server is ONLINE. \
+                    The server would be shutdown, updated, and brought back online."
+          ),
+          (false, true) => info!(
+            "Dry run: An update is available and the server is OFFLINE. \
+                    The server would be updated."
+          ),
+          (_, false) => info!("Dry run: No update is available. Nothing to do."),
+        }
+      } else if update_available {
+        debug!("Updating the installation!");
+        update_server()
+      }
+    }
+  }
+}
+
+fn update_server() {
+  // Shutdown the server if it's running
+  let server_was_running = server::is_running();
+  if server_was_running {
+    server::send_shutdown();
+    server::wait_for_exit();
```

Comment:
This seems to be a common usage. Add in something like a `blocking_stop` and rename `send_shutdown` to `send_shutdown_signal` to make it more clear?